### PR TITLE
Make nodes CIDR optional and allow it to be dynamic

### LINKED ISSUE #1582: Nodes CIDR should be optional
We are using the `nodes` CIDR in the `.spec.networking.nodes` field in order to properly set up the VPN connection between the kube-apiserver and the shoot worker nodes.
However, some providers don't need such a VPN connection because they only have one large network in which all their seed and shoot VMs are living together.
Consequently, we should make the field and the VPN configuration optional.

There is already https://github.com/gardener/vpn/pull/49 which makes it optional in our VPN component.

Softly depends on https://github.com/gardener/gardener-extensions/issues/407, I guess, in order to prevent invalid `Shoot` resources.

cc @stoyanr @mkoynov

### PR DESCRIPTION
**Pull request by rfranzke**
**What this PR does / why we need it**:
This PR incorporates two changes with respect to the CIDR for the nodes network for shoot clusters (the network that is used to run the shoot worker nodes):

1. The nodes CIDR can be optional now. This means that it won't be specified by the end-user in `core.gardener.cloud/v1{alpha,beta}1.Shoot.spec.networking.nodes`. The only thing that changes is that neither the `vpn-seed`s nor the `vpn-shoot` will be configured with the nodes CIDR (because there is none; this is possible as of gardener/vpn#49).
1. The nodes CIDR can be dynamic now. This means that it won't be specified by the end-user in `core.gardener.cloud/v1{alpha,beta}1.Shoot.spec.networking.nodes`. Instead, the infrastructure extension controller can provision the network dynamically during the reconciliation of the `Infrastructure` resource, and then write it into its new `.status.nodesCIDR` field. If present, Gardener will use this CIDR to configure both the `vpn-seed`s and the `vpn-shoot`. In case the user has already specified a nodes CIDR in the `Shoot` and the infrastructure extension controller writes a second CIDR into `.status.nodesCIDR`, then the latter one has the higher priority.

**Which issue(s) this PR fixes**:
Fixes #1582

**Special notes for your reviewer**:
* /cc @stoyanr @mkoynov - 1. will solve your use-case where a shoot cluster in your environment does not need to have a nodes network
* /cc @majst01 @Gerrit91 - 2. will solve your use-case where your infrastructure extension controller must dynamically provision a nodes network
* /cc @gardener/gardener-maintainers - for both we should consider implementing proper validation in the extension validator webhooks (gardener/gardener-extensions#407) - I have opened a PR for the already existing AWS validator: gardener/gardener-extensions#530 (cc @timuthy)

**Release note**:

```noteworthy developer
It is now possible for infrastructure extension controllers to dynamically provision a node network as part of the `Infrastructure` reconciliation. Please see [this document](https://github.com/gardener/gardener/blob/master/docs/extensions/infrastructure.md#dynamic-nodes-network-for-shoot-clusters) for further information.
```
```noteworthy developer
It is now possible to run Gardener in environments where no dedicated network for the worker nodes is required per shoot cluster (i.e., environments having a large network for all VMs of all clusters). In this case, the `Seed`s and the `Shoot`s do not need to specify a node CIDR. Gardener won’t configure the VPN components as it assumes network connectivity by default between the seed worker nodes and the shoot worker nodes.
```

### CODE REVIEW

File: charts/seed-controlplane/charts/kube-apiserver/values.yaml [hunk 1]
```diff
@@ -5,11 +5,9 @@ kubernetesVersion: 1.11.2
...
 probeCredentials: base64(user:pass)
 shootNetworks:
-  service: 10.0.1.0/24
-seedNetworks:
-  service: 10.0.0.0/24
-  pod: 192.168.0.0/1
-  node: 172.16.0.0/20
+  services: 10.0.1.0/24
+  pods: 192.168.0.0/1
+  nodes: 172.16.0.0/20
```
Comment (timuthy) [inline charts/seed-controlplane/charts/kube-apiserver/values.yaml]: Why not commented like done for `vpn-shoot` https://github.com/gardener/gardener/pull/1782/files?file-filters%5B%5D=.md&file-filters%5B%5D=.yaml#diff-25f9ee03324978298520e89310552fdfR15?

File: pkg/apis/garden/validation/validation_seed.go [hunk 2]
```diff
@@ -184,7 +186,9 @@ func ValidateSeedSpecUpdate(newSeedSpec, oldSeedSpec *garden.SeedSpec, fldPath *
 
 	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newSeedSpec.Networks.Pods, oldSeedSpec.Networks.Pods, fldPath.Child("networks", "pods"))...)
 	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newSeedSpec.Networks.Services, oldSeedSpec.Networks.Services, fldPath.Child("networks", "services"))...)
-	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newSeedSpec.Networks.Nodes, oldSeedSpec.Networks.Nodes, fldPath.Child("networks", "nodes"))...)
+	if newSeedSpec.Networks.Nodes != nil && oldSeedSpec.Networks.Nodes != nil {
```
Comment (timuthy) [inline pkg/apis/garden/validation/validation_seed.go]: This forbids to edit the CIDR directly but permits to remove (and later add) the field completely. So immutability isn't really preserved.

**Review by timuthy — CHANGES_REQUESTED** [review]
Nice PR. Just two comments and a general question: Have you considered to write back the `nodes` CIDR to the shoot resource when it was enforced by the infrastructure controller?

File: charts/seed-controlplane/charts/kube-apiserver/values.yaml (hunk 1 above)
Comment (rfranzke) [reply charts/seed-controlplane/charts/kube-apiserver/values.yaml]: Very nice catch, thanks, done!

File: pkg/apis/garden/validation/validation_seed.go (hunk 2 above)
Comment (rfranzke) [reply pkg/apis/garden/validation/validation_seed.go]: Same here, nice catch. I implemented some tests for the various scenarios.

**Comment by rfranzke** [comment]
@timuthy I am now updating the nodes CIDR in the shoot spec after reading it from the `Infrastructure`, makes sense! PTAL again.

File: pkg/operation/shoot/types.go [hunk 3]
```diff
@@ -46,6 +46,7 @@ type Shoot struct {
 	OperatingSystemConfigsMap map[string]OperatingSystemConfigs
 	Extensions                map[string]Extension
 	InfrastructureStatus      []byte
+	InfrastructureNodesCIDR   *string
```
Comment (timuthy) [inline pkg/operation/shoot/types.go]: Do you think this field/handling is still required after you write back the node CIDR to the shoot?

Comment (rfranzke) [reply pkg/operation/shoot/types.go]: Great feedback, @timuthy! Indeed, it's no longer required now. Also, I removed the check validating the disjointedness of the networks as it's done by the API server when trying to write the nodes CIDR into the shoot spec. Great!

**Review by timuthy — APPROVED** [review]
/lgtm

**Pull request merged**